<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<html>
<!-- This file was built using HtX, c. 1994-1997 by Richard M. Salter -->

<head>

<a name="@l10"><title>Introduction</title></a>
<basefont size=3>

</head>
<body BGCOLOR="#fffff5"         leftMargin="20">



<h3>Chapter: Introduction</h3> 

              <a href="help.html#@c10"><img src="./contents.gif" align=center alt="up" border=0>   &#160;<b>Contents</b></a>    &#160;  &#160;  &#160;  &#160;  &#160;  &#160;  &#160;  &#160;  &#160;  &#160;    <a href="help1.html#@l11"><img src="./forward.gif" align=center alt="forward" border=0>   &#160;<b>Installatio...</b></a><hr>   <i>Align</i> is an educational Java application for learning about and experimenting with pairwise sequence alignment algorithms used in biology. Align can be used to <ol ><li >Animate the Needleman-Wunsch and Smith-Waterman Algorithms; <li >Compare visual alignment with that produced by the algorithms; <li >Experiment with different scoring schemes </ol> <p> 
<i>Align</i> works with both amino acid and nucleotide sequences, using standard symbols. The program's scoring system accomodates both simple and matrix scoring rules. Standard scoring matrices, such as the Blosum and Pam matrices, may be imported and used in the algorithms.<p> 
     
<hr>
<address></address> <p>
</body>

</html>
